third_party/move/move-vm/runtime/src/storage/ty_depth_checker.rs

```diff
@@ -0,0 +1,730 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{module_traversal::TraversalContext, storage::loader::traits::StructDefinitionLoader};
+use move_binary_format::{
+    errors::{PartialVMError, PartialVMResult},
+    file_format::TypeParameterIndex,
+};
+use move_core_types::vm_status::StatusCode;
+use move_vm_metrics::{Timer, VM_TIMER};
+use move_vm_types::{
+    gas::GasMeter,
+    loaded_data::{
+        runtime_types::{DepthFormula, StructIdentifier, StructLayout, Type},
+        struct_name_indexing::StructNameIndex,
+    },
+};
+use std::{
+    cell::RefCell,
+    collections::{BTreeMap, HashMap, HashSet},
+    sync::Arc,
+};
+
+/// Checks depths for instantiated types.
+///
+/// For structs, stores a cache of formulas. The cache is used for performance (avoid repeated
+/// formula construction within a single transaction).
+///
+/// While a formula is being constructed, also checks for cycles between struct definitions. That
+/// is, cases like
+/// ```text
+/// struct A { b: B, }
+/// struct B { a: A, }
+/// ```
+/// are not allowed and constructing a formula for these types will fail.
+pub(crate) struct TypeDepthChecker<'a, T> {
```

Comment:
I'm assuming this file is just a refactor and there's existing test coverage?

Reply:
Yes? I combined code from interpreter.rs and from before into a single file (which logically should be in one place, adding gas meter + context, and adjusting the cyclic check).

We had 0 tests for this before actually... (the usual implicit testing via tests that pack and unpack structs/enums), so I added a few unit tests to check for cycles, and basic depths.

Reply:
A quick skimming through would be useful 😄